### ISSUE TITLE
Docker Agent can't use SSH for --base-url

### ISSUE THREAD

**Issue by thomasfrederikhoeck** [PrefectHQ/prefect#2770]
## Description
Docker Agent can't connect to remote Docker daemon with SSH. It results in
```
Traceback (most recent call last):
  File "C:\Users\tfh\Miniconda3\envs\prefect\lib\site-packages\docker\api\client.py", line 165, in __init__
    self._custom_adapter = SSHHTTPAdapter(
NameError: name 'SSHHTTPAdapter' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\tfh\Miniconda3\envs\prefect\Scripts\prefect-script.py", line 9, in <module>
    sys.exit(cli())
  File "C:\Users\tfh\Miniconda3\envs\prefect\lib\site-packages\click\core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "C:\Users\tfh\Miniconda3\envs\prefect\lib\site-packages\click\core.py", line 782, in main
    rv = self.invoke(ctx)
  File "C:\Users\tfh\Miniconda3\envs\prefect\lib\site-packages\click\core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "C:\Users\tfh\Miniconda3\envs\prefect\lib\site-packages\click\core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "C:\Users\tfh\Miniconda3\envs\prefect\lib\site-packages\click\core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Users\tfh\Miniconda3\envs\prefect\lib\site-packages\click\core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "C:\Users\tfh\Miniconda3\envs\prefect\lib\site-packages\click\decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "C:\Users\tfh\Miniconda3\envs\prefect\lib\site-packages\prefect\cli\agent.py", line 260, in start
    docker_interface=not no_docker_interface,
  File "C:\Users\tfh\Miniconda3\envs\prefect\lib\site-packages\prefect\agent\docker\agent.py", line 117, in __init__
    self.docker_client = self._get_docker_client()
  File "C:\Users\tfh\Miniconda3\envs\prefect\lib\site-packages\prefect\agent\docker\agent.py", line 142, in _get_docker_client
    return docker.APIClient(base_url=self.base_url, version="auto")
  File "C:\Users\tfh\Miniconda3\envs\prefect\lib\site-packages\docker\api\client.py", line 170, in __init__
    'Install paramiko package to enable ssh:// support'
docker.errors.DockerException: Install paramiko package to enable ssh:// support

```

## Expected Behavior
That I could connect to docker host.

## Reproduction
Simply run:
`prefect agent start docker --base-url ssh://mydockerhost`

## Environment
`
(prefect) λ prefect diagnostics
{
  "config_overrides": {
    "backend": true,
    "server": {
      "endpoint": true,
      "graphql": {
        "host": true
      },
      "host": true,
      "host_port": true,
      "port": true,
      "ui": {
        "endpoint": true,
        "graphql_url": true,
        "host": true,
        "host_port": true,
        "port": true
      }
    }
  },
  "env_vars": [],
  "system_information": {
    "platform": "Windows-10-10.0.18362-SP0",
    "prefect_version": "0.11.5",
    "python_version": "3.7.0"
  }
}
`

**Comment by joshmeek** [comment]
Thanks for opening @thomasfrederikhoeck! Will take a look at this. One thing, if you `pip install paramiko` then try running this again does it start working?

**Comment by thomasfrederikhoeck** [comment]
No luck. As far as I can tell it is something related to docker-compose: https://github.com/docker/compose/issues/6428

**Comment by thomasfrederikhoeck** [comment]
I was wandering if it was related to https://github.com/docker/compose/commit/7b82b2e8c721010b73f664e9d4657746a1fcd92b but I have
```
> docker-compose --version
docker-compose version 1.25.5, build 8a1c60f6
```

**Comment by thomasfrederikhoeck** [comment]
Can you replicate the error from running `prefect agent start docker --base-url ssh://mydockerhost`? @joshmeek

**Comment by joshmeek** [comment]
I do not yet currently have a remote version of the daemon running but if I try with some random IPs I don't get that original error you posted:

Invalid IP:
```
socket.gaierror: [Errno 8] nodename nor servname provided, or not known
```

Valid IP without open port:
```
paramiko.ssh_exception.NoValidConnectionsError: [Errno None] Unable to connect to port 22 on
```

**Comment by thomasfrederikhoeck** [comment]
Okay. I get that error no matter if I use a random host name (the one above) or my actual host name.

**Comment by thomasfrederikhoeck** [comment]
I have the following packages.
```
> pip list           
Package                 Version
----------------------- -------
appdirs                 1.4.3  
asn1crypto              1.3.0  
attrs                   19.3.0 
backcall                0.1.0  
bcrypt                  3.1.7  
bleach                  3.1.5  
brotlipy                0.7.0  
certifi                 2020.4.
cffi                    1.14.0 
chardet                 3.0.4  
click                   7.1.2  
cloudpickle             1.4.1  
colorama                0.4.3  
croniter                0.3.30 
cryptography            2.9.2  
cytoolz                 0.10.1 
dask                    2.18.1 
decorator               4.4.2  
defusedxml              0.6.0  
distributed             2.18.0 
docker                  4.2.1  
docker-pycreds          0.4.0  
entrypoints             0.3    
HeapDict                1.0.1  
idna                    2.9    
importlib-metadata      1.6.1  
ipykernel               5.3.0  
ipython                 7.15.0 
ipython-genutils        0.2.0  
jedi                    0.17.0 
Jinja2                  2.11.2 
json5                   0.9.0  
jsonschema              3.2.0  
jupyter-client          6.1.3  
jupyter-core            4.6.3  
jupyterlab              2.1.4  
jupyterlab-server       1.1.5  
MarkupSafe              1.1.1  
marshmallow             3.6.1  
marshmallow-oneofschema 2.0.1  
mistune                 0.8.4  
msgpack                 1.0.0  
mypy-extensions         0.4.3  
nbconvert               5.6.1  
nbformat                5.0.6  
notebook                6.0.3  
numpy                   1.18.5 
packaging               20.4   
pandas                  1.0.4  
pandocfilters           1.4.2  
paramiko                2.7.1  
parso                   0.7.0  
pendulum                2.1.0  
pickleshare             0.7.5  
pip                     20.1.1 
prefect                 0.11.5 
prometheus-client       0.8.0  
prompt-toolkit          3.0.5  
psutil                  5.7.0  
pycparser               2.20   
Pygments                2.6.1  
PyNaCl                  1.4.0  
pyOpenSSL               19.1.0 
pyparsing               2.4.7  
pyrsistent              0.16.0 
PySocks                 1.7.1  
python-box              4.2.3  
python-dateutil         2.8.1  
python-slugify          4.0.0  
pytz                    2020.1 
pytzdata                2019.3 
pywin32                 223    
pywinpty                0.5.7  
PyYAML                  5.3.1  
pyzmq                   19.0.1 
requests                2.23.0 
ruamel.yaml             0.16.6 
ruamel.yaml.clib        0.2.0  
Send2Trash              1.5.0  
setuptools              47.1.1.
six                     1.15.0 
sortedcontainers        2.2.1  
tabulate                0.8.7  
tblib                   1.6.0  
terminado               0.8.3  
testpath                0.4.4  
text-unidecode          1.3    
toml                    0.10.1 
toolz                   0.10.0 
tornado                 6.0.4  
traitlets               4.3.3  
Unidecode               1.1.1  
urllib3                 1.25.9 
wcwidth                 0.2.4  
webencodings            0.5.1  
websocket-client        0.57.0 
wheel                   0.34.2 
win-inet-pton           1.1.0  
wincertstore            0.2    
zict                    2.0.0  
zipp                    3.1.0  
```

**Comment by joshmeek** [comment]
What do you get when you run `docker version`? Wondering if it's a version issue, if not then it _could_ be windows related.

**Comment by thomasfrederikhoeck** [comment]
```
>docker version
Client: Docker Engine - Enterprise
 Version:           19.03.5
 API version:       1.40
 Go version:        go1.12.12
 Git commit:        2ee0c57608
 Built:             11/13/2019 08:00:16
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.3
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.10
  Git commit:       a872fc2f86
  Built:            Tue Oct  8 00:59:17 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.10
  GitCommit:        b34a5c8af56e510852c35414db4c1f4fa6172339
 runc:
  Version:          1.0.0-rc8+dev
  GitCommit:        3e425f80a8c931f88e6d94a8c831b9d5aa481657
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

```

**Comment by joshmeek** [comment]
Thanks, looks to be the same API version I have. @lauralorenz do you think you could test this on your windows machine?

**Comment by thomasfrederikhoeck** [comment]
It looks like it is a docker-compose related error as people are reporting this here https://github.com/docker/compose/issues/6336

**Comment by thomasfrederikhoeck** [comment]
Do you get the same error at @lauralorenz ? I guess a work around is to use Docker-In-Docker by building  a docker image with python, prefect and docker, with entrypoint `prefect agent start docker` and use `docker run -v /var/run/docker.sock:/var/run/docker.sock agent_starter`

**Comment by thomasfrederikhoeck** [comment]
I can further add that I also get a similar error when I'm trying to build the Flow using a remote Docker daemon over SSH as `flow.storage.build()` yields:

```
NameError                                 Traceback (most recent call last)
~\Miniconda3\envs\prefect\lib\site-packages\docker\api\client.py in __init__(self, base_url, version, timeout, tls, user_agent, num_pools, credstore_env)
    164             try:
--> 165                 self._custom_adapter = SSHHTTPAdapter(
    166                     base_url, timeout, pool_connections=num_pools

NameError: name 'SSHHTTPAdapter' is not defined

During handling of the above exception, another exception occurred:

DockerException                           Traceback (most recent call last)
<ipython-input-15-5ca469c77ed0> in <module>
----> 1 flow.storage.build()

~\Miniconda3\envs\prefect\lib\site-packages\prefect\environments\storage\docker.py in build(self, push)
    290 
    291         self.image_tag = self.image_tag or slugify(pendulum.now("utc").isoformat())
--> 292         self._build_image(push=push)
    293         return self
    294 

~\Miniconda3\envs\prefect\lib\site-packages\prefect\environments\storage\docker.py in _build_image(self, push)
    324             # Build the dockerfile
    325             if self.base_image and not self.local_image:
--> 326                 self.pull_image()
    327 
    328             dockerfile_path = self.create_dockerfile_object(directory=tempdir)

~\Miniconda3\envs\prefect\lib\site-packages\prefect\environments\storage\docker.py in pull_image(self)
    530             - InterruptedError: if either pulling the image fails
    531         """
--> 532         client = self._get_client()
    533 
    534         output = client.pull(self.base_image, stream=True, decode=True)

~\Miniconda3\envs\prefect\lib\site-packages\prefect\environments\storage\docker.py in _get_client(self)
    517 
    518         return docker.APIClient(
--> 519             base_url=self.base_url, version="auto", tls=self.tls_config
    520         )
    521 

~\Miniconda3\envs\prefect\lib\site-packages\docker\api\client.py in __init__(self, base_url, version, timeout, tls, user_agent, num_pools, credstore_env)
    168             except NameError:
    169                 raise DockerException(
--> 170                     'Install paramiko package to enable ssh:// support'
    171                 )
    172             self.mount('http+docker://ssh', self._custom_adapter)

DockerException: Install paramiko package to enable ssh:// support
```

**Comment by lauralorenz** [comment]
Hi @thomasfrederikhoeck, just wanted to let you know I could reproduce your issue on my windows machine and am taking a look. There is a lead internally on an anecdote of people having an issue if the python bindings for docker-compose is, due to the way it was installed (like with `brew` on macOS), are not on the conda env/ virtualenv path. Between that anecodote and your links you posted before I am going to do some digging on my side with my windows machine today.

**Comment by lauralorenz** [comment]
Oops I spoke too soon :) I could reproduce it because the conda environment I was in didn't have `paramiko` installed 😅 Once I `pip install paramiko`'d into the conda environment I was in I didn't get the error anymore. 

My situation FYI:
python version 3.7.7 [anaconda distribution]
paramiko 2.7.1 [from `pip list`]
docker version 19.03.8 (api version 1.40) [from `docker version`]
docker-compose version 1.25.5 (docker-py version 4.1.0) [from `docker-compose version`]
`docker` python package version 4.2.0 [from `pip list`]

To me it seems there must be some sort of python path issue between all of these dependencies for you somehow. @thomasfrederikhoeck in the python environment you ran the code for https://github.com/PrefectHQ/prefect/issues/2770#issuecomment-646524936, if you throw a `import paramiko` in the script or in the interpreter while you're doing that, does it import fine on its own?

**Comment by thomasfrederikhoeck** [comment]
So I tried running `import paramiko`
```
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-3-7bc9af46a7a8> in <module>
      1 from prefect import Flow, task
      2 from prefect.environments.storage import Docker
----> 3 import paramiko
      4 
      5 

~\Miniconda3\envs\prefect\lib\site-packages\paramiko\__init__.py in <module>
     20 import sys
     21 from paramiko._version import __version__, __version_info__
---> 22 from paramiko.transport import SecurityOptions, Transport
     23 from paramiko.client import (
     24     SSHClient,

~\Miniconda3\envs\prefect\lib\site-packages\paramiko\transport.py in <module>
    127 
    128 
--> 129 class Transport(threading.Thread, ClosingContextManager):
    130     """
    131     An SSH Transport attaches to a stream (usually a socket), negotiates an

~\Miniconda3\envs\prefect\lib\site-packages\paramiko\transport.py in Transport()
    188         "diffie-hellman-group1-sha1",
    189     )
--> 190     if KexCurve25519.is_available():
    191         _preferred_kex = ("curve25519-sha256@libssh.org",) + _preferred_kex
    192     _preferred_gsskex = (

~\Miniconda3\envs\prefect\lib\site-packages\paramiko\kex_curve25519.py in is_available(cls)
     28     def is_available(cls):
     29         try:
---> 30             X25519PrivateKey.generate()
     31         except UnsupportedAlgorithm:
     32             return False

~\Miniconda3\envs\prefect\lib\site-packages\cryptography\hazmat\primitives\asymmetric\x25519.py in generate(cls)
     36     @classmethod
     37     def generate(cls):
---> 38         from cryptography.hazmat.backends.openssl.backend import backend
     39         if not backend.x25519_supported():
     40             raise UnsupportedAlgorithm(

~\Miniconda3\envs\prefect\lib\site-packages\cryptography\hazmat\backends\openssl\__init__.py in <module>
      5 from __future__ import absolute_import, division, print_function
      6 
----> 7 from cryptography.hazmat.backends.openssl.backend import backend
      8 
      9 

~\Miniconda3\envs\prefect\lib\site-packages\cryptography\hazmat\backends\openssl\backend.py in <module>
     73     _CertificateSigningRequest, _RevokedCertificate
     74 )
---> 75 from cryptography.hazmat.bindings.openssl import binding
     76 from cryptography.hazmat.primitives import hashes, serialization
     77 from cryptography.hazmat.primitives.asymmetric import (

~\Miniconda3\envs\prefect\lib\site-packages\cryptography\hazmat\bindings\openssl\binding.py in <module>
     14 from cryptography import utils
     15 from cryptography.exceptions import InternalError
---> 16 from cryptography.hazmat.bindings._openssl import ffi, lib
     17 from cryptography.hazmat.bindings.openssl._conditional import CONDITIONAL_NAMES
     18 

ImportError: DLL load failed: The specified procedure could not be found.
```

This lead me to [Stack Overflow](https://stackoverflow.com/questions/48003901/paramiko-importerror-dll-load-failed-the-operating-system-cannot-run-1). Running `pip uninstall cryptography` and then `pip install cryptography` solved it. :tada: However I would like to now how it happen. According to [this](https://groups.google.com/forum/#!topic/scrapy-users/8N6V_OGUqtI) it might be something related to how `prefect` and therefore how `cryptography` is installed on Windows. How did you install prefect when you tried? @lauralorenz

**Comment by lauralorenz** [comment]
Glad to see that you got it worked out 🎉  Closing this now that we got it figured out.

To follow up on your questions: on my windows machine, I have `prefect` installed via `pip` from pypi, and then also from my local source using `pip`; out of curiosity I also installed it just now in another environment using `conda` instead (and followed the same package manager, respectively, for when I installed `paramiko`). I didn't experience the `cryptography` problems from any of the 3 situations.

All that being said, it is up to the user whether they install their dependencies using `pip` or `conda`. `cryptography` is a dependency of `paramiko`, not `prefect` (since `docker` chose not to make `paramiko` a hard dependency). If the speculation in that linked spacy google group is true and `cryptography` has issues with the Anaconda distro of python and `pip`, then it depends how the user install `paramiko` in particular if they will experience this issue with `cryptography`; and ultimately it is an upstream problem that is potentially best served by a ticket on the `cryptography` repo.
